provider/telegram.go

```diff
@@ -0,0 +1,395 @@
+		if err != nil {
+			t.Logf("failed to notify telegram peer: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Name of the handler
+func (t *TelegramHandler) Name() string { return t.ProviderName }
+
+// Default token lifetime. Changed in tests
+var tokenLifetime = time.Minute * 10
+
+// LoginHandler generates and verifies login requests
+func (t *TelegramHandler) LoginHandler(w http.ResponseWriter, r *http.Request) {
+	queryToken := r.URL.Query().Get("token")
+	if queryToken == "" {
+		// GET /login (No token supplied)
+		// Generate and send token
+		token, err := randToken()
+		if err != nil {
+			rest.SendErrorJSON(w, r, t.L, http.StatusInternalServerError, err, "failed to generate code")
+		}
+
+		t.mu.Lock()
+		t.authRequests[token] = authRequestInfo{
+			expires: time.Now().Add(tokenLifetime),
+		}
+		t.mu.Unlock()
+
+		fmt.Fprint(w, token)
+		return
+	}
+
+	// GET /login?token=blah
+	t.mu.Lock()
+	authRequest, ok := t.authRequests[queryToken]
+	t.mu.Unlock()
+
+	if !ok || time.Now().After(authRequest.expires) {
+		delete(t.authRequests, queryToken)
+		rest.SendErrorJSON(w, r, nil, http.StatusNotFound, nil, "request expired")
+		return
+	}
+
+	if !authRequest.confirmed {
+		rest.SendErrorJSON(w, r, nil, http.StatusNotFound, nil, "request not yet confirmed")
+		return
+	}
+
+	u := authtoken.User{
+		Name:    authRequest.user.Name,
+		ID:      t.ProviderName + "_" + authtoken.HashID(sha1.New(), fmt.Sprint(authRequest.user.ID)),
+		Picture: authRequest.user.Avatar,
+	}
+
+	u, err := setAvatar(t.AvatarSaver, u, &http.Client{Timeout: 5 * time.Second})
+	if err != nil {
+		rest.SendErrorJSON(w, r, t.L, http.StatusInternalServerError, err, "failed to save avatar to proxy")
+		return
+	}
+
+	claims := authtoken.Claims{
+		User: &u,
+		StandardClaims: jwt.StandardClaims{
+			Id:     queryToken,
+			Issuer: t.ProviderName,
+		},
+		SessionOnly: false, // TODO
+	}
+
+	if _, err := t.TokenService.Set(w, claims); err != nil {
+		rest.SendErrorJSON(w, r, t.L, http.StatusInternalServerError, err, "failed to set token")
+		return
+	}
+
+	rest.RenderJSON(w, r, claims.User)
+
+	// Delete request
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	delete(t.authRequests, queryToken)
+}
+
+// AuthHandler does nothing since we're don't have any callbacks
+func (t *TelegramHandler) AuthHandler(w http.ResponseWriter, r *http.Request) {}
+
+// LogoutHandler - GET /logout
+func (t *TelegramHandler) LogoutHandler(w http.ResponseWriter, r *http.Request) {
+	t.TokenService.Reset(w)
+}
+
+// tgAPI implements TelegramAPI
+type tgAPI struct {
+	logger.L
+	endpoint     string
+	token        string
+	updateOffset int
+}
+
+// NewTelegramAPI returns initialized TelegramAPI implementation
+func NewTelegramAPI(token string, l logger.L) TelegramAPI {
+	return &tgAPI{
+		L:        l,
+		endpoint: "https://api.telegram.org",
+		token:    token,
+	}
+}
+
+// GetUpdates fetches incoming updates
+func (t *tgAPI) GetUpdates(ctx context.Context) (*telegramUpdate, error) {
```

Comment:
`t` receiver a little bit confusing as this usually what we do in tests. Maybe call it `tg`? Also, the same t for two different structs in the same file makes it harder to follow. How about calling the first one `ti`?